### ISSUE TITLE
reuseaddr option is ignored on Windows

### ISSUE THREAD

**Issue by essen** [ninenines/ranch#216]
Sounds like `SO_REUSEADDR` on Linux is equivalent to `SO_REUSE_ADDR` + `SO_EXCLUSIVEADDR` on Windows. Might be worth patching OTP to support it.

Alternatively the new `socket` NIF should be made to work properly with it from scratch. But since it wants to be low-level perhaps it wants both options exposed instead?

* https://github.com/erlang/otp/blob/master/erts/emulator/drivers/common/inet_drv.c#L6478
* https://stackoverflow.com/questions/14388706/socket-options-so-reuseaddr-and-so-reuseport-how-do-they-differ-do-they-mean-t
* https://docs.microsoft.com/en-us/windows/desktop/WinSock/using-so-reuseaddr-and-so-exclusiveaddruse

**Comment by essen** [comment]
As long as nobody complains and the behavior is correct moving forward, which sounds like it is, then there is no problem.

**Comment by juhlig** [comment]
Ah, I remember that view on things of yours from bygone times :smile: Good old times :older_man:

**Comment by hazardfn** [comment]
@juhlig / @essen I believe we're experiencing a problem that may be tangentially related to what you're saying but I'm struggling to connect the dots, after updating to OTP 26 the following ranch option fails to work with the `socket` backend despite working on previous OTP versions (at least on my darwin dev machine, haven't tested linux):

```elixir
  @spec reuse_port() :: {:raw, any(), any(), any()} | nil
  defp reuse_port do
    case :os.type() do
      {:unix, :linux} -> {:raw, 1, 15, <<1::32-native>>}
      {:unix, :darwin} -> {:raw, 0xFFFF, 0x0200, <<1::32-native>>}
      _ -> nil
    end
  end
  ```
  
When switching back to the `inet` backend on OTP 26 this works as before. On at least OTP 24 and 25 this also works with the `socket` backend. I also found this so it could be a regression: https://github.com/erlang/otp/issues/5122

**Comment by essen** [comment]
Could be a regression or at least an unwanted side effect of the new developments. I would recommend writing a small snippet using `gen_tcp` directly to demonstrate the problem and opening a new ticket in OTP's repository.

**Comment by juhlig** [comment]
> fails to work with the `socket` backend

What does "fail to work" mean, ie what exactly happens or does not happen when?

**Comment by hazardfn** [comment]
> > fails to work with the `socket` backend
> 
> @hazardfn What does "fail to work" mean, ie what exactly happens or does not happen when?

Sorry for being unclear, basically SO_REUSEPORT is never set as intended and you end up with :eaddrinuse.

**Comment by juhlig** [comment]
Hm, does setting the `{reuseport, true}` instead of the `raw` option work for you, on 26?

**Comment by juhlig** [comment]
Oh wait, what was I thinking... That option does not work with the `socket` backend, that was what the ticket I opened at OTP was about 🤪

**Comment by juhlig** [comment]
> I strongly suspect this would also be the case for Linux but I don't have a testing machine to verify that.

You're right, same on Linux. I added it to https://github.com/erlang/otp/issues/7764#issuecomment-1807665811

**Comment by essen** [comment]
I will do a small 2.2 release soon. Is there anything to be done for that release in Ranch relating to this ticket?

**Comment by juhlig** [comment]
I don't think there is anything to add right now :man_shrugging:

FYI, status of erlang/otp#7764 is that the new options are not at all available for the socket backend right now (notices were put up in the docs), and regarding the raw option issue that @hazardfn pointed out, bmk just found out that they work as expected on 26.0 but not on 26.1 and will investigate.

**Comment by Maria-12648430** [comment]
Nothing from me either :woman_shrugging:

I mean, we _could_ do some fiddling with the default and disallowed listen options, like for Windows setting `{reuseaddr, false}` on OTP<26 and `{reuseaddr, true}, {reuseport, true}, {exclusiveaddruse, true}` on OTP>=26 (and disallow changing them) (unless the socket backend is used, where those options are not supported yet). Stuff like that. Not sure if it's worth the trouble.

**Comment by juhlig** [comment]
FYI: https://github.com/erlang/otp/issues/7764#issuecomment-1842815863

**Comment by essen** [comment]
You could send a PR with a test case for this and it should be succeeding in about a week when CI recompiles OTP master.
